### ISSUE TITLE
[BUG][File Composition] allowOnlySpecifiedSelectors.fileExport `false` doesn't work

### ISSUE THREAD

**Issue by FredericEspiau** [Igorkowalski94/eslint-plugin-project-structure#21]
1. Make sure you have the latest version before reporting a bug.

**Describe the bug**

The next `type` defined after the latest `type` in the configuration is reported as an error even though `allowOnlySpecifiedSelectors.fileExport` is `false`.

Configuration

```
// @ts-check

import { createFileComposition } from "eslint-plugin-project-structure";

export const fileCompositionConfig = createFileComposition({
  filesRules: [
    {
      filePattern: "./src/domain/**/*.ts",
      allowOnlySpecifiedSelectors: {
        fileExport: false,
        fileRoot: false,
        nestedSelectors: false,
      },
      rules: [
        {
          selector: ["type"],
          scope: "fileExport",
          format: "Id",
          positionIndex: 0,
        },
        {
          selector: ["variableExpression"],
          scope: "fileExport",
          format: "Id",
          positionIndex: 1,
        },
        {
          selector: ["type"],
          scope: "fileExport",
          format: "Props",
          positionIndex: 2,
        },
      ],
    },
  ],
});
```

File

```
export type Id = IdRaw<"Batch">;
export const Id = IdRaw("Batch");

export type Props = any;

export type GetBatch = (id: Id) => any; // 🔥 Invalid type name, allowed formats = Id, Props 🔥
```

**Expected behavior**

No error should happen for `GetBatch` as not only the specified selectors are supposed to match.

**Additional context**

Thanks for all the hard work.

**Comment by Igorkowalski94** [comment]
In this case, the error is correct. You have only specified an exported type with the name Props or Id in your file. If you also want to allow dynamic names, e.g., {PascalCase}, you should write it like this:

```jsoc
// @ts-check

import { createFileComposition } from "eslint-plugin-project-structure";

export const fileCompositionConfig = createFileComposition({
  filesRules: [
    {
      filePattern: "./src/domain/**/*.ts",
      allowOnlySpecifiedSelectors: {
        fileExport: false,
        fileRoot: false,
        nestedSelectors: false,
      },
      rules: [
        {
          selector: ["type"],
          scope: "fileExport",
          format: "Id",
          positionIndex: 0,
        },
        {
          selector: ["variableExpression"],
          scope: "fileExport",
          format: "Id",
          positionIndex: 1,
        },
        {
          selector: ["type"],
          scope: "fileExport",
          format: "Props",
          positionIndex: 2,
        },
        {
          selector: ["type"],
          scope: "fileExport",
          format: "{PascalCase}",
        },
      ],
    },
  ],
});
```

**Comment by FredericEspiau** [comment]
Thanks I guess I didn't understand how it works. I wonder if it's confusing to other people as well though? Closing as it works

**Comment by Igorkowalski94** [comment]
If you have any suggestions to make it clearer, feel free to share.

**Comment by FredericEspiau** [comment]
Well I feel the current way it works is too restrictive as I wanted the extension to enforce some rules in my file, not the whole content for my file for a given selector.

Here are some ideas I just got:

```
{
  selector: ["interface"],
  scope: "fileExport",
  format: "NoOtherInterface",
}
```

As there is no `position` configuration, this means `NoOtherInterface` should be the only exported `interface` in the file

---

```
{
  selector: ["interface"],
  scope: "fileExport",
  format: "FirstInterface",
  position: {
    index: 0,
  },
}
```

This means that `FirstInterface` should be the first exported `interface` in the file, so you could have other exported `interface` after

---

A more complicated configuration:

```
{
    selector: ["type"],
    scope: "fileExport",
    format: "TypeName",
    position: [
      {
        scope: ["fileExport", "fileRoot"]
        index: 0,
      },
      {
        scope: "fileExport",
        last: true,
      },
    ],
},
```

This configuration means that `TypeName`

- should be the first exported `type` selector in the file, both for exported and non-exported types
- should be the last export `type` selector in the file, so you could have more types in the file but only `TypeName` should be exported

---

Another kind of complicated configuration:

```
{
  selector: ["interface"],
  scope: "fileExport",
  format: "NoOtherInterface",
},
{
    name: "FirstSelector", // property to make references to other selectors
    selector: ["type"],
    scope: "fileExport",
    format: "FirstSelector",
    position: {
      index: 0,
    },
},
{
  selector: ["type"],
  scope: "fileExport",
  format: "InTheMiddle",
  position: { // could be implicit as there is already an exported type selector with a position before
    after: "FirstSelector",
  }
},
{
    selector: ["variable"],
    scope: "fileExport",
    format: "LastVariable",
    position: {
      selector: ["variable"],
      last: true
    }
},
{
    selector: ["type"],
    scope: "fileExport",
    format: "LastSelector",
    position: {
      last: true
    }
},
```

Means that:

- `FirstSelector` should be the first selector in the file
- `NoOtherInterface` should be the only `interface` in the file as it has no `position` configuration
- `InTheMiddle` should be anywhere in the file but after `FirstSelector` (and obviously before `LastSelector`), so you could have other exported types between `InTheMiddle` and `LastSelector`
- `LastVariable` should be the last `variable` in the file
- `LastSelector` should be the last selector in the file

**Comment by Igorkowalski94** [comment]
Maybe first show me the ideal file you want to achieve along with the correct order, etc. I will write the configuration for it, and you will have a complete comparison.

**Comment by Igorkowalski94** [comment]
Currently, the rules without positionIndex will always be last. I can think about improving the positionIndex, for example, -1 will indicate the last position, -2 the second to last, etc.

**Comment by Igorkowalski94** [comment]
I added the ability to provide a negative `positionIndex`, allowing you to determine the order from the end.
 [v3.5.0](https://github.com/Igorkowalski94/eslint-plugin-project-structure/releases/tag/v3.5.0)

**Comment by Igorkowalski94** [comment]
Additionally, `allowOnlySpecifiedSelectors` is used to prohibit the use of selectors that don't have their own rules. If you've already defined a rule for a given selector, such as `positionOrder` for a unique type, and you don't want to define a naming rule for other types, you can write it like this:

```mjs
// @ts-check

import { createFileComposition } from "eslint-plugin-project-structure";

export const fileCompositionConfig = createFileComposition({
  filesRules: [
    {
      filePattern: "./src/domain/**/*.ts",
      allowOnlySpecifiedSelectors: {
        fileExport: false,
        fileRoot: false,
        nestedSelectors: false,
      },
      rules: [
        {
          selector: ["type"],
          scope: "fileExport",
          format: "Id",
          positionIndex: 0,
        },
        {
          selector: ["variableExpression"],
          scope: "fileExport",
          format: "Id",
          positionIndex: 1,
        },
        {
          selector: ["type"],
          scope: "fileExport",
          format: "Props",
          positionIndex: 2,
        },
        {
          selector: "type",
          scope: "fileExport",
          format: "*",
        },
      ],
    },
  ],
});
```
